File: javascript/0003-longest-substring-without-repeating-characters.js
/**
 * 3. Longest Substring Without Repeating Characters
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 **/

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  if (s.length <= 1) return s.length;

  let l = 0,
    r = 0,
    max = 1,
    now = 1,
    letters = {};
  letters[s[l]] = true;

  while (++r < s.length) {
    while (letters[s[r]] !== undefined && letters[s[r]]) {
      letters[s[l++]] = false;
      now--;
    }
    letters[s[r]] = true;
    now++;
    max = Math.max(max, now);
  }

  return max;
};

File: swift/0003-longest-substring-without-repeating-characters.swift
/**
 * 3. Longest Substring Without Repeating Characters
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 **/

class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        let characters: [String] = Array(s).map { String($0) }
        var hashmap: [String: Bool] = [:]
        var l = 0, r = 0
        var ans = 0

        while (r < characters.count) {
            let ch = characters[r]
            while hashmap[ch] != nil {
                hashmap[characters[l]] = nil
                l += 1
            }
            hashmap[ch] = true

            ans = max(ans, r - l + 1)

            r += 1
        }

        return ans
    }
}